Add foreign keys to data object tables

BlackbitDevs opened the pull request:
From time to time we have the problem that there is an item in the `object_store_*` table but not in `objects`. As there is a unique data field it is then neither possible to create a new item with the same value, nor to find the existing one (because of INNER JOINS on the `objects` table. I do not know the reason, perhaps there is a transaction missing somewhere but anyway, it would be cleaner if the data which belongs together would be explicitly connected with a foreign key. Then such situations could not occur anymore, no matter how one of the datasets belonging to the same object got removed / not created.

Commit:
add foreign keys to data store tables

Commit:
add foreign keys to data store tables

Commit:
add foreign keys to query table

brusch commented:
@BlackbitNeueMedien I agree, that totally makes sense! 
Just some more thoughts on it: 
- we should implement it consistently across all object related tables, so including localized fields, object bricks and field collections
- I think we can remove `ON UPDATE` as updating the ID is anyway not supported
- we should create migrations for existing tables
- we should remove resulting obsolete code 

For now I'd focus on data objects, then we could and should extend it at least to all other entities (Documents & Assets).

Commit:
add foreign keys to data object tables

Commit:
add foreign keys to data object tables

Commit:
add foreign keys to data object tables

Commit:
delete object-related data in other tables via foreign keys

Commit:
alter referencing object id columns to be INT(11) UNSIGNED

Commit:
shorten foreign key name if longer than 64 characters

Commit:
shorten foreign key name if longer than 64 characters

BlackbitDevs commented:
Is now implemented for all data object-related tables.
I removed the "ON UPDATE CASCADE" (as @brusch mentioned)
Wrote a migration for existing tables.
Removed now unnecessary deletion of object-related data as this now gets automatically deleted via foreign keys.

brusch commented on bundles/CoreBundle/Migrations/Version20211018104331.php:
```diff
@@ -0,0 +1,110 @@
+
+                $fkName = AbstractDao::getForeignKeyName($table, $objectIdColumn);
+                if (!$tableSchema->hasForeignKey($fkName)) {
+                    if($tableSchema->getColumn($objectIdColumn)->getCustomSchemaOption('unsigned') === false) {
```
```suggestion
                    $column = $tableSchema->getColumn($objectIdColumn);
                    if(!$column->hasCustomSchemaOption('unsigned') || $column->getCustomSchemaOption('unsigned') === false) {
```
Otherwise this would result in `Warning: Undefined array key "unsigned" `

brusch commented on bundles/CoreBundle/Migrations/Version20211018104331.php:
```diff
@@ -0,0 +1,110 @@
+            'object_classificationstore_groups_'.$class->getId() => 'o_id',
+            'object_classificationstore_data_'.$class->getId() => 'o_id',
+            'object_metadata_'.$class->getId() => 'o_id',
+            'object_localized_data_'.$class->getId() => 'ooo_id'
```
`object_localized_query_*` tables are missing here.

BlackbitDevs replied:
Fixed.

brusch commented on bundles/CoreBundle/Migrations/Version20211018104331.php:
```diff
@@ -0,0 +1,110 @@
+                        $tableSchema->changeColumn($objectIdColumn, ['unsigned' => true]);
+                    }
+
+                    $tableSchema->addForeignKeyConstraint('objects', [$objectIdColumn], ['o_id'], ['ON DELETE CASCADE'], $fkName);
```
```suggestion
                    $tableSchema->addForeignKeyConstraint('objects', [$objectIdColumn], ['o_id'], ['onDelete' => 'CASCADE'], $fkName);
```
The current code however adds the FK, but with default action `RESTRICT`.

brusch commented on bundles/CoreBundle/Migrations/Version20211018104331.php:
```diff
@@ -0,0 +1,110 @@
+                    continue;
+                }
+
+                $fkName = 'fk_'.$table.'__'.$objectIdColumn;
```
```suggestion
                $fkName = AbstractDao::getForeignKeyName($table, $objectIdColumn);
```

brusch commented on bundles/CoreBundle/Migrations/Version20211018104331.php:
```diff
@@ -0,0 +1,110 @@
+
+        $brickList = new DataObject\Objectbrick\Definition\Listing();
+        foreach ($brickList->load() as $brickDefinition) {
+            $foreignKeys['object_brick_query_'.$brickDefinition->getKey().'_'.$class->getId()] = 'o_id';
```
`object_brick_store_*` tables are missing here.

BlackbitDevs replied:
Fixed.

brusch commented on bundles/CoreBundle/Migrations/Version20211018104331.php:
```diff
@@ -0,0 +1,110 @@
+    private function getForeignKeys(DataObject\ClassDefinition $class): array
+    {
+        $foreignKeys = [
+            'object_query_'.$class->getId() => 'oo_id',
```
Additionally we should also add `object_url_slugs.objectId` , this also needs to be considered in `install.sql`.

BlackbitDevs replied:
Fixed.

brusch commented on models/DataObject/Concrete/Dao.php:
```diff
@@ -419,19 +419,6 @@ public function saveChildData()
-        if ($this->model->getClass()) {
-            foreach ($this->model->getClass()->getFieldDefinitions() as $fd) {
-                if ($fd instanceof CustomResourcePersistingInterface) {
-                    $fd->delete($this->model);
```
This needs to stay, because there could be data-types which are e.g. not using the database at all but e.g. Redis.

BlackbitDevs replied:
Fixed.

brusch commented on models/DataObject/ClassDefinition/Dao.php:
```diff
@@ -127,32 +127,35 @@ public function update()
               INDEX `forward_lookup` (`src_id`, `ownertype`, `ownername`, `position`),
-              INDEX `reverse_lookup` (`dest_id`, `type`)
+              INDEX `reverse_lookup` (`dest_id`, `type`),
+			  CONSTRAINT `".self::getForeignKeyName($objectDatastoreTableRelation, 'src_id')."` FOREIGN KEY (`src_id`) REFERENCES objects (`o_id`) ON DELETE CASCADE
```
We could do that for `dest_id` column as well, right?

BlackbitDevs replied:
I think this is not possible because `dest_id` could also refer to assets or documents. 

Edit: This seems to be named `polymorphic foreign key` and it is [possible](https://stackoverflow.com/questions/2002985/mysql-conditional-foreign-key-constraints#67708114) but this seems to get too complicated imho. Also I do not know if all supported DBMS vendors support such *generated columns*. What do you think, @brusch ?

brusch replied:
@BlackbitNeueMedien you're right, I'd leave it as it is for now. Thanks a lot for your changes 😊

Commit:
Update bundles/CoreBundle/Migrations/Version20211018104331.php

Co-authored-by: Bernhard Rusch <brusch@users.noreply.github.com>

Commit:
Update bundles/CoreBundle/Migrations/Version20211018104331.php

Co-authored-by: Bernhard Rusch <brusch@users.noreply.github.com>

Commit:
Update bundles/CoreBundle/Migrations/Version20211018104331.php

Co-authored-by: Bernhard Rusch <brusch@users.noreply.github.com>

Commit:
add foreign keys to data object tables

Commit:
add foreign keys to data object tables

brusch commented on lib/Model/Dao/AbstractDao.php:
```diff
@@ -95,4 +95,13 @@ public function resetValidTableColumnsCache($table)
+    public static function getForeignKeyName($table, $column) {
+        $fkName = 'fk_'.$table.'__'.$column;
+        if (strlen($fkName) > 64) {
+            $fkName = 'fk_'.md5($fkName);
```
```suggestion
            $fkName = substr($fkName, 0, 55) . '_' . hash('crc32', $fkName);
```
I think keeping a prefix here would be nice and `crc32` should be enough randomness.

brusch committed:
Update lib/Model/Dao/AbstractDao.php

brusch commented:
Well done 👍 Thanks a lot!
